src/lib/prov/tpm2/tpm2_crypto_backend/tpm2_crypto_backend.cpp

```diff
@@ -566,16 +573,66 @@ static TSS2_RC get_ecdh_point(TPM2B_PUBLIC* key,
                               BYTE* out_buffer,
                               size_t* out_size,
                               void* userdata) {
+   #if defined(BOTAN_HAS_TPM2_ECC_ADAPTER)
+   return thunk([&] {
+      BOTAN_UNUSED(userdata);
+      BOTAN_ASSERT_NONNULL(key);
+      BOTAN_ASSERT_NOMSG(key->publicArea.type == TPM2_ALG_ECC);
+
+      const auto curve_name = Botan::TPM2::curve_id_tss2_to_botan(key->publicArea.parameters.eccDetail.curveID);
+      const auto curve_order_byte_size =
+         Botan::TPM2::curve_id_order_byte_size(key->publicArea.parameters.eccDetail.curveID);
+      if(!curve_name || !curve_order_byte_size) {
+         return TSS2_ESYS_RC_NOT_IMPLEMENTED;  // ECC curve not supported
+      }
+
+      const auto domain = Botan::EC_Group::from_name(curve_name.value());
+
+      BOTAN_ASSERT_NOMSG(curve_order_byte_size.value() == domain.get_p_bytes());
+
+      Botan::AutoSeeded_RNG rng;  //TODO: What RNG to use?
+
+      // 1: Generate ephemeral key
+      const Botan::ECDH_PrivateKey eph_key(rng, domain);
+
+      // Get public key coordinates
+      const Botan::PointGFp& eph_pub_point = eph_key.public_point();
+      const auto& bn_x = eph_pub_point.get_affine_x();
+      const auto& bn_y = eph_pub_point.get_affine_y();
+
+      // Write public ephemeral affine coordinates to TPMS_ECC_POINT (Q)
+      if(bn_x.bytes() != curve_order_byte_size || bn_y.bytes() != curve_order_byte_size) {
+         // TODO: Is actually correct to check bytes() > curve_order_byte_size?
+         return TSS2_ESYS_RC_GENERAL_FAILURE;
+      }
+
+      // Serialize affine coordinates into TPM2B_ECC_PARAMETER with Big Endian encoding
+      Botan::TPM2::copy_into<TPM2B_ECC_PARAMETER>(Q->x, bn_x.serialize(curve_order_byte_size.value()));
+      Botan::TPM2::copy_into<TPM2B_ECC_PARAMETER>(Q->y, bn_y.serialize(curve_order_byte_size.value()));
+
+      // 2: Get TPM public key
+      const auto tpm_pub_inputs = Botan::TPM2::ecc_pubkey_from_tss2_public(key);
+      const auto tpm_sw_pubkey = Botan::ECDH_PublicKey(tpm_pub_inputs.first, tpm_pub_inputs.second);
+
+      // 3: ECDH Key Agreement
+      Botan::PK_Key_Agreement ecdh(eph_key, rng, "Raw" /*No KDF used here*/);
+      const auto shared_secret = ecdh.derive_key(0 /*Ignored for raw KDF*/, tpm_sw_pubkey.public_value()).bits_of();
+
+      Botan::TPM2::copy_into(*Z, shared_secret);
+
+      size_t offset = 0;
+      const auto r = Tss2_MU_TPMS_ECC_POINT_Marshal(Q, out_buffer, max_out_size, &offset);
+      if(r != TSS2_RC_SUCCESS) {
+         return r;
+      }
```

Comment:
Couldn't we use the `check_rc` helper here? The thunk will catch the exception and return the error code from there. Not as efficient for sure, but its an error case. (No strong opinion on that)